docs/writing-tests.md

```diff
@@ -0,0 +1,60 @@
+
+- **Second critical functional check**. A very similar test exists at a lower level but the functionality is so absolutely critical we test it again:
+
+  - [Good example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/frontend.feature#L18) - undeniably a critical page for GOV.UK ([lower level test exists](https://github.com/alphagov/frontend/blob/a40ba988c58dd882d841a6f1d82d5b894939fdc5/test/integration/help_test.rb#L18))
+  - [Good example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/frontend.feature#L59) - one of the most critical features of GOV.UK ([lower level test exists](https://github.com/alphagov/router/blob/96d5912cec4c8052c2f9f71d257f2add2787b65d/integration_tests/redirect_test.go?L23#L12))
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L29) - checks an arbitrary, non-critical page ([lower level test exists](https://github.com/alphagov/collections/blob/cc94525a05b2651fee621b588e530aca8e1f77e4/features/viewing_browse.feature))
+  - [Bad example](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/finder_frontend.feature#L45) - missing a lower level test in the app itself ([example test to copy](https://github.com/alphagov/finder-frontend/blob/d70a01903e2813719e7d5adbf79fbcb1d49335fb/spec/controllers/email_alert_subscriptions_controller_spec.rb#L363))
+
+Do not duplicate tests. For example, ["Email signup from an organisation home page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L47) duplicates ["Email signup from a taxon page"](https://github.com/alphagov/smokey/blob/aaa0522666d49db116b939276f0e0969357e3ed4/features/collections.feature#L53): both cover data transfer with Email Alert API.
+
+Do not add tests that can be implemented in other ways. We want this to be the tip of the [testing pyramid](https://martinfowler.com/bliki/TestPyramid.html), not a [testing ice cream cone](http://saeedgatson.com/the-software-testing-ice-cream-cone/). Instead of adding a test here, you can:
+
+- Add or extend [health checks for application infrastructure](https://github.com/alphagov/govuk_app_config/blob/main/docs/healthchecks.md). For example, drafting a document in a publishing app involves connecting to infrastructure like a database or AWS S3.
+
+- Add [contract tests](https://docs.publishing.service.gov.uk/manual/pact-broker.html) to cover a chain of APIs. For example, subscribing to an email involves communication with Email Alert API and Content Store, which can be tested via GDS API Adapters.
+
+- Add unit or integration tests to cover in-app behaviour. For example, the way each document format is rendered can be tested in the frontend app that does the rendering.
+
```

Comment:
We have a few tests that try to measure latency - which is better served by monitoring application metrics rather than a test. Using a test is guaranteed to be flakey - as we never (and shouldn't) expect 100% of requests to meet our requirements. 

Wonder if we should add that here as guidance?

Reply:
Yeah, I guess just adding guidance here could help prevent regression and someone adding a new test for latency.

Reply:
Sorry, I don't think I was clear in my comment: we already have a "bad example" about that sort of test, which I think of as guidance. Do you think it's worth being even more explicit about performance tests?